### DISCUSSION TITLE
A11y discussion: Combobox / MultiSelect Gray onHover contrast

### DISCUSSION THREAD

**Discussion by andreancardona** [carbon-design-system/carbon#7929] [Help] [answered]
## Background

It was brought to our attention by an Accessibility expert that our current design when a user hovers over one of the options but it is not selected is too similar to an option that is selected. 

## Testing / Replicate

To see this, type 2 and then Enter (Option 2 is selected). Now type down arrow. Option 3 currently has focus, but it looks almost the same as Option 2 except for the checkmark.

<img width="387" alt="109088732-3ab13f80-76cd-11eb-83ab-f7b2ee54cfe6" src="https://user-images.githubusercontent.com/32720851/109543614-c0ddd500-7a83-11eb-8c40-aa38d49aa273.png">

## Ask / Discussion: 

- I know this is a change that would update multiple components which is why I wanted to open it up for discussion. 
- Would love design feedback on this and any other thoughts

## Potential Solutions

1. Get rid of the on hover color change completely  -- this is a completely valid solution that can be seen in this thread here: https://github.com/carbon-design-system/carbon/pull/7777#issuecomment-786702375

2. Create more distinction between the Hover color background and the Selected background

**Comment by thyhmdo** [comment]
- I personally don't have a big concern about the color since I can see the Checkmark and the option inside the input field. 
- If we choose to create the distinction between the Hover color background and the Selected background: 
    - Material uses a different hover color (Behavior section) https://material.io/design/interaction/states.html#selected
    - Should we also be concerned about the hover state in the data table as well? https://carbon-components-react.netlify.app/?path=/story/components-datatable-selection--usage

**Comment by aagonzales** [comment]
1. I don't think we would want to remove the hover. I think that would be making the experience worse. 
2. Hover is consistent across our components so I don't think that would be the color to change. If we were to change anything it would be the selected state. 

**Other thoughts:**
- Originally we didn't include the checkmark for selected state. Adding it was our solution to help distinguish between selected and hover. 
- The selected option also appears in the dropdown field itself so I don't think there is much confusion around what is selected with that and the checkmark.
- The hover is noted by the background change and the cursor the user is controlling. 
- For keyboard "active" I can see a where this might start to be confusing. Originally I think design thought the blue focus border would move each item but the focus stays on the list box so we could look into how to better account for this.

**Reply by aagonzales** [reply]
Now I'm wondering if you @andreancardona mistyped. Did you mean  "Get rid of the on ~hover~ **selected** color change completely"

That is what I'm seeing in the browser selects. 

Chrome
![image](https://user-images.githubusercontent.com/11670886/109551031-07d8c400-7a9e-11eb-815d-48bc4660569a.png)

Safari
![image](https://user-images.githubusercontent.com/11670886/109551130-28a11980-7a9e-11eb-995b-0df22204490f.png)

Firefox
![image](https://user-images.githubusercontent.com/11670886/109551211-42426100-7a9e-11eb-9489-1e2ce1584fc9.png)

**Comment by aagonzales** [comment]
@andreancardona is this actually violating an accessibility rule? Why isn't the checkmark a sufficient indicator for selection?

**Reply by laurenmrice** [reply]
I had the same question in the accessibility [combobox PR](https://github.com/carbon-design-system/carbon/pull/7777#issuecomment-786269776). As far as I know, revising this design could help low vision users, but I don't think it is actually violating an accessibility rule. Going to tag @dakahn to weigh in more on this when you have some time.

**Reply by aagonzales** [reply]
Ok I read through that issue a little better. Seems like some people are [saying](https://github.com/carbon-design-system/carbon/pull/7777#issuecomment-786403458) its better to have a stronger emphasis on the hovered/focus item than the selected item. So I guess they are implying that the checkbox is sufficient and doesn't need a background to help differentiate it.

**Reply by laurenmrice** [reply]
@andreancardona Is the bigger problem actually that the keyboard hover needs a visual indicator that passes 3:1 ?

**Comment by kingtraceyj** [comment]
so it sounds like it's more of a keyboard user issue, right? so when a user uses the arrow keys to switch an item, the hover is close to the selected. so maybe it's more of a problem when using a keyboard and trying to select something that's right next to the thing that is selected.

the other issue it seems mike gower has is the contrast with the background. i see that our dropdown does have enough contrast because the menu is a select state but combobox does not because of what is selected:
![image](https://user-images.githubusercontent.com/20582179/109591586-666c6500-7ad3-11eb-9e27-695cc21a7375.png)



essentially mike gower is looking for something like this really quick mock up (the active state gets a little weird with combobox since the field is actually selected but included a version like dropdown even though it's not technically correct — just to look at):
![image](https://user-images.githubusercontent.com/20582179/109591327-ef36d100-7ad2-11eb-8728-690ee4cdc9ac.png)

**Comment by aagonzales** [comment]
These are all the other components I could think of that use a selected darker background. So right now dropdown is following the system. If we change dropdown to not include a background change would we want change it everywhere?

![image](https://user-images.githubusercontent.com/11670886/109706019-6e291980-7b66-11eb-90ca-53c69821c2b9.png)

**Reply by aagonzales** [reply]
Honestly, if is nots breaking an actual accessibility rule I think we should keep it to match the system. If we think there needs to be a better distinction between the hover and selected then I think we need to look at that across these components.

**Reply by aagonzales** [reply]
What other components would need this:
- Structured list: NO - uses focus/tab for navigation
- Data table: not sure, how does one navigate to the text in a row?
- Icon button: NO - uses focus/tab for navigation
- Left nav: NO - uses focus/tab for navigation

**Comment by jnm2377** [comment]
Maybe auditing selected/hover states in all our components and seeing if there are better options is an effort we should focus on in our next major release. In my opinion, the hover/selected color difference has always felt weird to me bc it's so subtle, esp when navigating via keyboard, it looks so similar. Also, when an item is selected and it's being hovered on, there is not background change, which also doesn't seem right. On selectable data table, we still have the selected row change bg color when it's hovered on. 

 I think, it is also important to note that some a11y testing needs to be done manually, which means that even if the Accessibility Checker doesn't throw and error, it doesn't mean that the component is 100% accessible inherently. I was trying to test this against different color blindness, but also wanna do more research as to how this affects keyboard users. 

My first thought, is that something that might help is to have the focus on the menu item when navigating via keyboard instead of the input like we do for OverflowMenu. I thought we did this, but maybe I was wrong or it changed. If it did change, I'd love to know more about why (if it was related to an a11y thing or a design change). I'd expect, that when I'm arrowing up or down through the items, they would have a focus border, but they don't. I think this would be more effective for keyboard users than just the bg hover change. As far as for mouse users, I'm not sure what would be a good fix since I wouldn't want a focus border when actually hovering.

**Reply by aagonzales** [reply]
>Also, when an item is selected and it's being hovered on, there is not background change, which also doesn't seem right.

There should be a background hover change, we have a token `hover-selected-ui`. You can see it in effect with selected data table rows. There was a bug with dropdown that we've fixed this release.

**Reply by aagonzales** [reply]
>My first thought, is that something that might help is to have the focus on the menu item when navigating via keyboard instead of the input like we do for OverflowMenu.

This was the original design intent but it was changed for a11y reasons. I would like to know more about this as well. Why does the focus stay on the list-box and not move to each item? I honestly get a little confused when focus get trapped in an area and you use arrow keys instead of tabs to navigate. What is that called if not focus? Active?

**Reply by aagonzales** [reply]
>In my opinion, the hover/selected color difference has always felt weird to me bc it's so subtle, esp when navigating via keyboard, it looks so similar. 

The background change is a second indicator. All selected items (except ghost button) have a check mark or selected bar to indicate so I don't see why its an accessibility problem. I don't see anywhere where its a rule that hover and selected backgrounds have to have 3:1 contrast difference. In fact that would be pretty strong if it did. We'd have to go up to Gray60 to get that level of contrast. So it seems like an aesthetic choices.

That's why I wonder if its just the style of the keyboard navigation state that is the true problem not hover and selected colors.

**Comment by dakahn** [comment]
Just catching up on this conversation now. I'll try and provide some guidance if I can.

I guess the first thing worth noting is that multiple selection in a listbox from an accessibility standpoint is generally inadvisable and a group of check boxes is _always_ preferable if possible. 

With that said what we have here is both a UX problem and an accessibility violation with our keyboard users getting the short shrift. 

As a quick aside -- something that may help is to stop considering MultiSelect as one singular component and think of it more as a compound component. With two distinct sub-components: a button and a listbox. The user's focus moves fluidly between these two components and selections are made in one that are represented in another.

So with the button (input) focused via <kbd>tab</kbd> the user then can use either <kbd>space</kbd> or <kbd>enter</kbd> to open the listbox. When the listbox is opened focus moves from the button to the listbox. At which point the user can use <kbd>up</kbd> or <kbd>down</kbd> arrow keys to navigate _within_ the listbox and make their selection(s).

![image](https://user-images.githubusercontent.com/40970507/110400306-09296400-803d-11eb-8f25-d7ee4d950110.png)
So here we see a _satisfactory_ amount of contrast between the active/inactive elements, but let's make a few selections. 

So now with the listbox opened and selections made we see this
![image](https://user-images.githubusercontent.com/40970507/110401302-ee57ef00-803e-11eb-8b04-2ec8523c5755.png)
So the first question that comes to mind is _within the listbox_ where is my active list item? It is not perceivable. 

In fact if the keyboard user has closed and then reopened the listbox they'll have to hit <kbd>down arrow</kbd> 4 times before an active list item can be discerned as seen here
![image](https://user-images.githubusercontent.com/40970507/110401756-ca48dd80-803f-11eb-97d6-338ab46ce394.png)

Aside from a poor experience for our keyboard users we have a contrast violation. The selected state and the active state need a 3:1 contrast as per the [guidelines for UI Distinguishability](https://www.ibm.com/able/toolkit/design/visual/color-and-contrast/) in the IBM Accessibility Requirements which state,  "any visual enhancement that indicates a state change needs to maintain a 3:1 contrast ratio against the meaningful adjacent colors".

**Reply by aagonzales** [reply]
>The selected state and the active state need a 3:1 contrast as per the guidelines for UI Distinguishability in the IBM Accessibility Requirements which state, "any visual enhancement that indicates a state change needs to maintain a 3:1 contrast ratio against the meaningful adjacent colors".

Exactly, it not necessary a "we need to change the background color" problem its a we need to give the active state a 3:1 indicator. We can definitely look into this. In [V9](https://v9.carbondesignsystem.com/components/dropdown/code) we had the active state identifiable by appling an underline to the active text. We can look into some other possible visual indicators that's not a background change. 

![image](https://user-images.githubusercontent.com/11670886/110494856-fa38c500-80c1-11eb-8f14-f5b28ad53318.png)

If we only did it with a background color change this is what it would have to look like to get 3:1 which is not optimal.

![image](https://user-images.githubusercontent.com/11670886/110496161-391b4a80-80c3-11eb-85da-30ced17ee346.png)

**Reply by aagonzales** [reply]
Thanks for explaining all of that @dakahn !

**Reply by dakahn** [reply]
I can't react with emojis for some reason but 😸 I'm so psyched I could help out

**Reply by laurenmrice** [reply]
Great explanation @dakahn ! Thank you 💯

**Comment by aagonzales** [comment]
Here are some possible visual indicator options for the new `active` state. (this is different from hover)

![image](https://user-images.githubusercontent.com/11670886/110531265-5bc05a00-80e9-11eb-8b75-343c90fb516f.png)
![image](https://user-images.githubusercontent.com/11670886/110531313-6844b280-80e9-11eb-8906-afec7dfb87e6.png)
![image](https://user-images.githubusercontent.com/11670886/110531361-7692ce80-80e9-11eb-9537-5e100e3018c4.png)

**Reply by laurenmrice** [reply]
Thanks for making these explorations! ⭐️ 🙌🏻

Thoughts on some of these:
- Underline of the text in each option could almost look like it is a link, and it is not.
- Inverse background color of an unselected option could be confused as an option that is selected because the tag is also an inverse color.
- The bottom third section with decorative elements could make the menu a little busy.

-----

I do like the outline versions 👍🏻. All blue or all black outlines looks clean to me. The all blue might resonate more with people since its keyboard navigation.

**Comment by aagonzales** [comment]
Question: Can `active` and `hover` be styled differently? @dakahn 

I was just nothing in firefox search bar it has a more subtle `hover` for mouse and then a much stronger keyboard `active` state. 

![Mar-09-2021 13-00-34](https://user-images.githubusercontent.com/11670886/110516369-7db0e100-80d7-11eb-800d-9ef04136f680.gif)

**Reply by dakahn** [reply]
They absolutely can! As long as they're separately distinguishable 👍🏽

**Reply by aagonzales** [reply]
Oh nice! That actually makes things a lot easy there. And I think then we can have a bolder active state to help with keyboard nav.

**Comment by aagonzales** [comment, marked as answer]
# Proposed solution for active state in list box
@laurenmrice and I discussed in slack, we like this option. It keeps with our focus style, even though it's not technically focused. I will show it in Thursday's design crit to confirm with the rest of the team. 

![image](https://user-images.githubusercontent.com/11670886/110536041-fec7a280-80ee-11eb-9046-4147c0e8ef8c.png)

**Reply by aagonzales** [reply]
@dakahn @andreancardona @jnm2377 do you think this would satisfy the accessibility requirement?

**Reply by jnm2377** [reply]
For keyboard users, yes. That's a lot more accessible than what we currently have. @dakahn is the focus on the whole menu still needed? I feel like it would make more sense to just have the focus on the active item rather than the whole menu.

@aagonzales Would this border be applied on hover, as well, for mouse users? Otherwise, the low contrast would still be an issue if we keep the current hover color.

**Reply by aagonzales** [reply]
No, the hover would stay as is without the blue border. I think its less of an issue with mouse hover because the user will be pointing at each item themselves so I don't think it gets lost anywhere. Also I don't think hover has to pass 3:1 otherwise basically the entire internet would do it wrong.

**Reply by dakahn** [reply]
That would be a really great improvement! Would be way more usable and accessible for non mouse users. 👍🏽
